File: express/mapper/url.mapper.js
let toEntity = (dto, shortCode, qr) => ({
  originalUrl: dto.originalUrl,
  shortCode,
  qrCode: qr,
});

let toResponse = (entity, baseUrl) => ({
  originalUrl: entity.originalUrl,
  shortUrl: baseUrl + entity.shortCode,
  qrCodeBase64: entity.qrCode.toString("base64"),
});

export { toEntity, toResponse };

File: spring boot/src/main/java/com/urlshortnerproj/tinyurlClone/Mapper/UrlMapper.java
package com.urlshortnerproj.tinyurlClone.Mapper;

import com.urlshortnerproj.tinyurlClone.DTO.UrlRequestDTO;
import com.urlshortnerproj.tinyurlClone.Entity.UrlMapping;
import com.urlshortnerproj.tinyurlClone.DTO.UrlResponseDTO;
import java.util.Base64;
public class UrlMapper {
    public static UrlMapping toEntity(UrlRequestDTO reqDto,String shortCode,byte[] qrCode,UrlMapping entity){
            entity.setOriginalUrl(reqDto.getOriginalUrl());
            entity.setShortUrl(shortCode);
            entity.setQrCode(qrCode);
            return entity;

    }

    public static UrlResponseDTO toResponse(
            UrlMapping entity,
            String baseUrl
    ) {
        return new UrlResponseDTO(
                entity.getOriginalUrl(),
                baseUrl + entity.getShortUrl(),
                Base64.getEncoder().encodeToString(entity.getQrCode())
        );
    }
}
